In [11]:
# import necessary modules and algorithms
import pandas as pd
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.linear_model import LinearRegression

In [12]:
for i in range(1, 18):
    print(str(i).zfill(2))

01
02
03
04
05
06
07
08
09
10
11
12
13
14
15
16
17


In [13]:
# load all patient files (1-53) and store in dictionary
patients = {}
for m in range(1, 54):
    patients[f"patient{m}"] = pd.read_csv(
        f'bidmc_{str(m).zfill(2)}_Numerics.csv',
        na_values='NaN'
    )

In [14]:
x = 1
len(patients[f'patient{x}']) # returns the amount of rows in the profile

481

In [15]:
# add an id column for each patient (1-53)
for id in range(1, 54):
    patients[f'patient{id}']['id'] = [id]*len(patients[f'patient{id}'])

In [26]:
# Extract age and gender data from text files

patient_age = {}
patient_gender_M = {}
patient_gender_F = {}

for x in range(1, 54):
    with open (f'bidmc_{str(x).zfill(2)}_Fix.txt') as pt:
        for rec in pt:
            if rec[0:3] == 'Age':
                if not rec[5:7] == 'Na':
                    patient_age[f'patient{x}'] = rec[5:7]
                else:
                    patient_age[f'patient{x}'] = 'NaN'
            if rec[0:3] == 'Gen':
                if rec[8] == 'M':
                    patient_gender_M[f'patient{x}'] = 1
                    patient_gender_F[f'patient{x}'] = 0
                elif rec[8] == 'F':
                    patient_gender_F[f'patient{x}'] = 1
                    patient_gender_M[f'patient{x}'] = 0
                else:
                    print(f'patient{x}')
                    patient_gender_F[f'patient{x}'] = 'NaN'
                    patient_gender_M[f'patient{x}'] = 'NaN'
                

# create new columns for age and gender and assign them to each patient
for dat in range(1,54):
    patients[f'patient{dat}']['age'] = patient_age[f'patient{dat}']
    patients[f'patient{dat}']['gender_M'] = patient_gender_M[f'patient{dat}']
    patients[f'patient{dat}']['gender_F'] = patient_gender_F[f'patient{dat}']

patient18


In [25]:
patients['patient18']

Time [s]   HR   PULSE   RESP   SpO2  id  age gender_M gender_F
0           0   91      89     18     98  18  NaN      NaN      NaN
1           1   91      89     18     98  18  NaN      NaN      NaN
2           2   91      89     18     98  18  NaN      NaN      NaN
3           3   91      90     18     99  18  NaN      NaN      NaN
4           4   91      90     18     99  18  NaN      NaN      NaN
..        ...  ...     ...    ...    ...  ..  ...      ...      ...
476       476   90      90     18     98  18  NaN      NaN      NaN
477       477   90      90     18     98  18  NaN      NaN      NaN
478       478   90      89     18     98  18  NaN      NaN      NaN
479       479   90      89     18     98  18  NaN      NaN      NaN
480       480   90      89     18     98  18  NaN      NaN      NaN

[481 rows x 9 columns]

In [27]:
# How to access column from dictionary of dataframes
patients["patient1"][' HR'] 


0      94
1      94
2      94
3      92
4      93
       ..
476    90
477    90
478    90
479    90
480    90
Name:  HR, Length: 481, dtype: int64

In [28]:
# rename column names for all patient files
for patient in patients:
    patients[patient].rename({
        "Time [s]": "time",
        " HR":"hr",
        " PULSE":"pulse",
        " RESP":"resp",
        " SpO2":"spo2"
    }, axis=1, inplace=True)

In [29]:
patients['patient1']

time  hr  pulse  resp  spo2  id age  gender_M  gender_F
0       0  94   93.0    25  97.0   1  88         1         0
1       1  94   93.0    25  97.0   1  88         1         0
2       2  94   93.0    25  97.0   1  88         1         0
3       3  92   93.0    26  97.0   1  88         1         0
4       4  93   93.0    26  97.0   1  88         1         0
..    ...  ..    ...   ...   ...  ..  ..       ...       ...
476   476  90   90.0    21  96.0   1  88         1         0
477   477  90   90.0    21  96.0   1  88         1         0
478   478  90   90.0    21  96.0   1  88         1         0
479   479  90   90.0    20  96.0   1  88         1         0
480   480  90   90.0    20  96.0   1  88         1         0

[481 rows x 9 columns]

In [10]:
# m = 1

# patient_01 = pd.read_csv(f'bidmc_0{m}_Numerics.csv', na_values='NaN')
# patient_02 = pd.read_csv('bidmc_02_Numerics.csv')
# patient_03 = pd.read_csv('bidmc_03_Numerics.csv')

# patients = [patient_01, patient_02, patient_03]

# p_01 = patient_01.rename({
#     "Time [s]": "time",
#     " HR":"hr",
#     " PULSE":"pulse",
#     " RESP":"resp",
#     " SpO2":"spo2"
# }, axis=1)

# p_02 = patient_02.rename({
#     "Time [s]": "time",
#     " HR":"hr",
#     " PULSE":"pulse",
#     " RESP":"resp",
#     " SpO2":"spo2"
# }, axis=1)

# p_03 = patient_03.rename({
#     "Time [s]": "time",
#     " HR":"hr",
#     " PULSE":"pulse",
#     " RESP":"resp",
#     " SpO2":"spo2"
# }, axis=1, inplace=True)

# for patient in d:
#     patient.rename({
#         "Time [s]": "time",
#         " HR":"hr",
#         " PULSE":"pulse",
#         " RESP":"resp",
#         " SpO2":"spo2"
#     }, axis=1, inplace=True)

In [52]:
#identify which patients have missing information
import numpy as np
patient_features = ['time', 'hr', 'pulse', 'resp', 'spo2', 'age', 'gender_M', 'gender_F']
indicator = 1
for patient in patients:
    patients[patient].replace('NaN', np.nan, inplace=True)
    
    for feature in patient_features:
        if patients[patient][feature].isna().sum() > 0:
            print(f"patient{indicator}")
            print(patients[patient].isna().sum())
            break
    indicator = indicator + 1

patient1
time         0
hr           0
pulse       13
resp         0
spo2        13
id           0
age          0
gender_M     0
gender_F     0
dtype: int64
patient5
time        0
hr          0
pulse       6
resp        0
spo2        6
id          0
age         0
gender_M    0
gender_F    0
dtype: int64
patient13
time          0
hr            0
pulse         0
resp        146
spo2          0
id            0
age           0
gender_M      0
gender_F      0
dtype: int64
patient15
time        0
hr          0
pulse       0
resp        3
spo2        0
id          0
age         0
gender_M    0
gender_F    0
dtype: int64
patient18
time          0
hr            0
pulse         0
resp          0
spo2          0
id            0
age         481
gender_M    481
gender_F    481
dtype: int64
patient19
time          0
hr            0
pulse       102
resp          4
spo2        102
id            0
age           0
gender_M      0
gender_F      0
dtype: int64
patient25
time        0
hr          0
pulse  

## List of patients with missing data
 patient 1 (13 pulse/spo2) |
 patient 13 (146 resp) |
 patient 15 (3 resp) |
 patient 18 (481 age/gender) |
 patient 19 (102 pulse/spo2 and 4 resp) |
 patient 27 (4 hr), patient 25 (1 pulse/spo2) |
 patient 44 (6 pulse/spo2) |

In [12]:
# remove rows where there are missing pulse values
for patient in patients:
    patients[patient].dropna(how='any', subset=['pulse'], inplace=True)

In [13]:
patients['patient44'].isnull().sum()

time     0
hr       0
pulse    0
resp     0
spo2     0
id       0
dtype: int64

In [14]:
# make new list of patient profiles concatinated together
f = 1
new_list= []
# patients[f'patient{f}']
for patient in patients:
    new_list.append(patients[patient])

In [22]:
new_list

[     time  hr  pulse  resp  spo2  id
 0       0  94   93.0    25  97.0   1
 1       1  94   93.0    25  97.0   1
 2       2  94   93.0    25  97.0   1
 3       3  92   93.0    26  97.0   1
 4       4  93   93.0    26  97.0   1
 ..    ...  ..    ...   ...   ...  ..
 476   476  90   90.0    21  96.0   1
 477   477  90   90.0    21  96.0   1
 478   478  90   90.0    21  96.0   1
 479   479  90   90.0    20  96.0   1
 480   480  90   90.0    20  96.0   1
 
 [468 rows x 6 columns],
      time  hr  pulse  resp  spo2  id
 0       0  93     92    16   100   2
 1       1  93     93    16   100   2
 2       2  93     93    16   100   2
 3       3  94     94    16   100   2
 4       4  94     94    16   100   2
 ..    ...  ..    ...   ...   ...  ..
 476   476  89     92    15   100   2
 477   477  89     92    15   100   2
 478   478  89     91    15   100   2
 479   479  89     91    15   100   2
 480   480  89     90    15   100   2
 
 [481 rows x 6 columns],
      time  hr  pulse  resp  spo2 

In [15]:
# Combine all patient profiles into one
result = pd.concat(new_list)

In [23]:
result

time    hr  pulse  resp   spo2  id
0       0  94.0   93.0  25.0   97.0   1
1       1  94.0   93.0  25.0   97.0   1
2       2  94.0   93.0  25.0   97.0   1
3       3  92.0   93.0  26.0   97.0   1
4       4  93.0   93.0  26.0   97.0   1
..    ...   ...    ...   ...    ...  ..
476   476  92.0   95.0  18.0  100.0  53
477   477  90.0   95.0  19.0  100.0  53
478   478  97.0   95.0  19.0  100.0  53
479   479  99.0   95.0  19.0  100.0  53
480   480  94.0   95.0  19.0  100.0  53

[25208 rows x 6 columns]

In [16]:
result.isnull().sum()

time       0
hr         4
pulse      0
resp     153
spo2       0
id         0
dtype: int64

In [17]:
# drop all the rows with missing resp values
result.dropna(how='any', subset=['resp'], inplace=True)

In [18]:
# drop all rows with missing hr values
result.dropna(how='any', subset=['hr'], inplace=True)

In [19]:
# Assign independent and dependent variables
X_01 = result.drop(columns=['resp'])
y_01 = result['resp']


In [20]:
X_01.shape

(25208, 5)

Apply algorithms to datasets and score using cross validation method

In [21]:


ranfor = RandomForestRegressor(random_state=1)
#ranfor.fit(X_01, y_01)

from sklearn.model_selection import cross_val_score, StratifiedKFold
cross_val_score(ranfor, X_01, y_01, cv=StratifiedKFold(shuffle=True,
                                                       n_splits=4,
                                                       random_state=1)).mean()

C:\Users\BioGuest\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:666: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=4.
  warnings.warn(("The least populated class in y has only %d"


KeyboardInterrupt: 

In [ ]:
linreg = LinearRegression()

cross_val_score(linreg, X_01, y_01, cv=StratifiedKFold(shuffle=True,
                                                      n_splits=8,
                                                      random_state=1)).mean()

In [ ]:
gradbst = GradientBoostingRegressor()

from sklearn.model_selection import cross_val_score, StratifiedKFold
cross_val_score(gradbst, X_01, y_01, cv=StratifiedKFold(shuffle=True,
                                              n_splits=8,
                                              random_state=1)).mean()

In [ ]:
from sklearn.svm import SVR

svr = SVR()

cross_val_score(svr, X_01, y_01, cv=StratifiedKFold(shuffle=True,
                                              n_splits=8,
                                              random_state=1)).mean()

Some optimisation on the above algorithms could be done (GridSearchCV/RandomizedSearchCV) to improve performance, but first the data would need to be split into train and test sets.

In [ ]:
y_01.shape